packages/react-aria-components/stories/Table.stories.tsx

```diff
@@ -392,3 +393,134 @@ const MyCheckbox = ({children, ...props}: CheckboxProps) => {
     </Checkbox>
   );
 };
+
+const MyTableLoader = () => {
+  return (
+    <UNSTABLE_TableLoader>
+      {({isTableEmpty}) => (
+        <span>
+          {isTableEmpty ? 'Loading spinner' : 'Load more spinner'}
+        </span>
+      )}
+    </UNSTABLE_TableLoader>
+  );
+};
+
+const TableLoadingBodyWrapper = (args: {isLoading: boolean}) => {
+  return (
+    <Table aria-label="Files">
+      <TableHeader columns={columns}>
+        {(column) => (
+          <Column isRowHeader={column.isRowHeader}>{column.name}</Column>
+        )}
+      </TableHeader>
+      <MyTableBody rows={rows} isLoading={args.isLoading}>
+        {(item) => (
+          <Row columns={columns}>
+            {(column) => {
+              return <Cell>{item[column.id]}</Cell>;
+            }}
+          </Row>
+        )}
+      </MyTableBody>
+    </Table>
+  );
+};
+
+function MyTableBody(props) {
+  let {rows, children, isLoading, ...otherProps} = props;
+  return (
+    <TableBody {...otherProps}>
+      <Collection items={rows}>
+        {children}
+      </Collection>
+      {isLoading && <MyTableLoader />}
+    </TableBody>
+  );
+}
+
+export const TableLoadingBodyWrapperStory = {
+  render: TableLoadingBodyWrapper,
+  args: {
+    isLoading: false
+  },
+  name: 'Table loading, table body wrapper with collection'
+};
+
+const TableLoadingRowRenderWrapper = (args: {isLoading: boolean}) => {
+  return (
+    <Table aria-label="Files">
+      <TableHeader columns={columns}>
+        {(column) => (
+          <Column isRowHeader={column.isRowHeader}>{column.name}</Column>
+        )}
+      </TableHeader>
+      <TableBody items={rows} dependencies={[args.isLoading]}>
+        {(item) => (
+          <MyRow columns={columns} isLoading={item.id === 4 && args.isLoading}>
+            {(column) => {
+              return <Cell>{item[column.id]}</Cell>;
+            }}
+          </MyRow>
+        )}
+      </TableBody>
+    </Table>
+  );
+};
+
+function MyRow(props) {
+  return (
+    <>
+      {/* Note that all the props are propagated from MyRow to Row, ensuring the id propagates */}
+      <Row {...props} />
+      {props.isLoading && <MyTableLoader />}
+    </>
+  );
+}
+
+export const TableLoadingRowRenderWrapperStory = {
+  render: TableLoadingRowRenderWrapper,
+  args: {
+    isLoading: false
+  },
+  name: 'Table loading, row renderer wrapper and dep array'
+};
+
+
+// TODO: when writing the docs, prefer this pattern in the docs, establish renderEmptyState as something distinct from empty + loading
+// (aka users should use empty state for if their table is empty and not loading, but should render the TableLoader for empty + loading and for has items + loading)
```

Comment:
Oh. What was the reason for not using `renderEmptyState` for loading? Not sure which one is better I guess, but seems like there was a bit of extra complexity to handle the empty state with TableLoader.

Reply:
Mainly felt confusing that renderEmptyState would then have mixed responsibilities for rendering the true empty state and the initial loading state